drgn_tools/meminfo.py

```diff
@@ -0,0 +1,755 @@
+
+# Using five bits for the type means that the maximum number of
+# swapcache pages is 27 bits * on 32-bit-pgoff_t architectures.
+_MAX_SWAPFILES_SHIFT = 5
+# NUMA node memory migration support
+_SWP_DEVICE_NUM = 2
+# NUMA node memory migration support
+_SWP_MIGRATION_NUM = 2
+# Handling of hardware poisoned pages with memory corruption.
+_SWP_HWPOISON_NUM = 1
+
+# |MAX_SWAPFILES| defines the maximum number of swaptypes.
+_MAX_SWAPFILES = (
+    (1 << _MAX_SWAPFILES_SHIFT)
+    - _SWP_DEVICE_NUM
+    - _SWP_MIGRATION_NUM
+    - _SWP_HWPOISON_NUM
+)
+
+# The index for unknown stats ID for node stats and zone stats.
+UNKNOWN_GLOBAL_STAT_ID = -1
+
+
+def print_val_kb(text: str, num: int) -> None:
+    """
+    Produce the formatted output that matches the output in /proc/meminfo.
+
+    param text: The output item's name, e.g, "MemTotal"
+    param num: The output item's value
+    """
+    if num < 0:
+        return
+    text += ":"
+    num_kb = num << (_PAGE_SHIFT - 10)
+    print("%s%s kB" % (text.ljust(16, " "), str(num_kb).rjust(8, " ")))
+
+
+def parse_machine_info(prog: Program) -> None:
+    """
+    Parse the system's information to determine architecture-specific parameters.
+    This function must be called before parsing vmcore's memory statistics.
+    """
+    global _IS_X86_64, _IS_AARCH_64, _UEK_VERSION
+    global _MAX_NUM_NODES, _MAX_NUM_ZONES
+    global _PAGE_SIZE, _PAGE_SHIFT, _PMD_SHIFT, _HPAGE_SHIFT
+    global _HPAGE_PMD_SHIFT, _HPAGE_PMD_ORDER, _HPAGE_PMD_NR
+    global _VMALLOC_END, _VMALLOC_START, _VMALLOC_SPACE
+
+    uts = get_uts(prog)
+    # Set UEK version
+    if "4.14.35" in uts["release"]:
+        _UEK_VERSION = 5
+    elif "5.4.17" in uts["release"]:
+        _UEK_VERSION = 6
+    else:
+        _UEK_VERSION = 7
+
+    # Determine page size
+    _PAGE_SIZE = prog.constant("PAGE_SIZE").value_()
+    _PAGE_SHIFT = prog.constant("PAGE_SHIFT").value_()
+
+    # Set arch
+    if "aarch64" in uts["machine"]:
+        _IS_AARCH_64 = True
+
+        # include/linux/numa.h
+        _NODES_SHIFT = 2
+
+        # arch/arm64/include/asm/pgtable-hwdef.h
+        n = 2
+        _PMD_SHIFT = (_PAGE_SHIFT - 3) * (4 - (n)) + 3
+        n = 1
+        PGDIR_SHIFT = (_PAGE_SHIFT - 3) * (4 - (n)) + 3
+
+        # include/asm-generic/pgtable-nopud.h
+        P4D_SHIFT = PGDIR_SHIFT
+        PUD_SHIFT = P4D_SHIFT
+        PUD_SIZE = 1 << PUD_SHIFT
+
+        # arch/arm64/include/asm/pgtable.h
+        _HPAGE_SHIFT = _PMD_SHIFT
+
+        SZ_64K = 0x00010000
+        SZ_128M = 0x08000000
+        VA_BITS = 48
+        VA_START = 0xFFFFFFFFFFFFFFFF - (1 << VA_BITS) + 1
+        PAGE_OFFSET = 0xFFFFFFFFFFFFFFFF - (1 << (VA_BITS - 1)) + 1
+        STRUCT_PAGE_MAX_SHIFT = 6
+        KASAN_SHADOW_SIZE = 0
+        MODULES_VADDR = VA_START + KASAN_SHADOW_SIZE
+        MODULES_VSIZE = SZ_128M
+        MODULES_END = MODULES_VADDR + MODULES_VSIZE
+        VMEMMAP_SIZE = 1 << (VA_BITS - _PAGE_SHIFT - 1 + STRUCT_PAGE_MAX_SHIFT)
```

Comment:
Using the upper case for these variables is confusing. I see that you're using `_` prefix for all your private global variables (that's great!). But since these are locals, the `ALL_CAPS` naming convention is confusing if you're just reading this function.

I know you're making these variables comparable to what you'd find in the corresponding header file, but I'd prefer if you made them lower case so it's easier to read, and also easier to see what is the actual final output of these computations.

Reply:
Updated accordingly.